api/smb.proto

```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
+
+package api;
+
+import "errors.proto";
+
+service SmbCSIProxyService {
+    // MountSMBShare mounts a remote share over SMB on the host at the requested global staging path.
+    rpc MountSMBShare(MountSMBShareRequest) returns (MountSMBShareResponse) {}
+}
+
+message MountSMBShareRequest {
+    // A remote SMB share to mount
+    // All unicode characters allowed in SMB server name specifications are
+    // permitted except for restrictions below
+    //
+    // Restrictions:
+    // SMB share specified in the format: \\server-name\sharename, \\server.fqdn\sharename or \\a.b.c.d\sharename
+    // If not an IP address, server-name has to be a valid DNS name.
+    // UNC specifications to local paths or prefix: \\?\ is not allowed.
+    // Characters: + [ ] " / : ; | < > , ? * = $ are not allowed.
+    string remote_share = 1;
+
+    // Local path in the host to stage the SMB share.
+    // All special characters allowed by Windows in path names will be allowed
+    // except for restrictions noted below. For details, please check:
+    // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+    //
+    // Restrictions:
+    // If given an absolute path, it needs to be under kubelet-csi-plugins-path,
+    // and specified with drive letter prefix: "X:\".
+    // If given a relative path, it will be understood as being relative to
+    // kubelet-csi-plugins-path.
+    // Needs to already exist on host.
```

Comment:
There are two ways to use the `New-SmbGlobalMapping` command for mounting SMB shares on the container host to prepare for mounting them as container volumes:

1) `New-SmbGlobalMapping -RemotePath "\\$ServerName\$ShareName" -LocalPath X: -Credential cred` . This seems to be what `MountSMBShareRequest` is targeting, but in this case it might make more sense to just accept a drive letter here (which does not need to exist on the host I think?), and then use the other proxy APIs to create any child directories.

2) `New-SmbGlobalMapping -RemotePath "\\$ServerName\$ShareName" -Credential $cred` (no `-LocalPath`) followed by a symlink to e.g. `C:\smbshare`. This is the approach used by the existing kubernetes [SMB mount package](https://github.com/kubernetes/kubernetes/blob/2036ee5ddeac235c0c39c705ad22cfa19bad309e/pkg/util/mount/mount_windows.go#L109). `MountSMBShareRequest` should allow this (or a separate request type should be created) since in our experience we've encountered strange permission errors when mounting SMB shares created using approach 1) to containers, whereas shares created using approach 2) can be mounted to containers as expected.

cc @almos98

Reply:
I think approach 2) is also what we're targeting here; does 1) allow for some use cases that 2) doesn't?

Reply:
FYI the limitation with approach 1) is described at https://github.com/moby/moby/issues/37863.

I guess I'm confused by these comments for this method:
```
    // If given an absolute path, it needs to be under kubelet-csi-plugins-path,
    // and specified with drive letter prefix: "X:\".
    // If given a relative path, it will be understood as being relative to
    // kubelet-csi-plugins-path.
    // Needs to already exist on host.
```

The X:\ letter and "needs to already exist on host" are confusing. For approach 2) this method could either just create the mapping and then let the driver call a separate method to create the symlink, or this method could do both but I think the path would need to *not* already exist on the host (and the drive letter would typically be C:\).

Reply:
Fair points, changing :)